File: Project 1/UQ_8_8.m
clear 
close all

tf = 5;
dt = 0.1;
t_vals = 0:dt:tf;
S0 = 900; R0 = 0; I0 = 100;  N=1000;
S_gamma0 = 0; I_gamma0 = 0; R_gamma0 = 0;
S_k0 = 0; I_k0 = 0; R_k0 = 0;
S_delta0 = 0; I_delta0 = 0; R_delta0 = 0;
S_r0 = 0; I_r0 = 0; R_r0 = 0;

gamma = 0.2; k = 0.1; delta = 0.15; r = 0.6;
params=[gamma k delta r];
Y0 = [S0; I0; R0; S_gamma0; I_gamma0; R_gamma0; S_k0; I_k0; R_k0; S_delta0; I_delta0; R_delta0;  S_r0; I_r0; R_r0];
% Integrate the coupled system using ode45.m.  There are now N(p+1) coupled
% differential equations comprised of the state and sensitivity equations
ode_options = odeset('RelTol',1e-8);
[~,Y] = ode45(@SIR_sens,t_vals,Y0,ode_options,params);
% Extract the states and sensitivities. 
S = Y(:,1);             I = Y(:,2);                  R = Y(:,3);
S_gamma_sen = Y(:,4);   I_gamma_sen = Y(:,5);        R_gamma_sen = Y(:,6);
S_k_sen = Y(:,7);       I_k_sen = Y(:,8);            R_k_sen = Y(:,9);
S_delta_sen = Y(:,10);  I_delta_sen = Y(:,11);       R_delta_sen = Y(:,12);
S_r_sen = Y(:,13);      I_r_sen = Y(:,14);           R_r_sen = Y(:,15);
% Compute the sensitivities using complex-step derivative approximations.  This
% requires p=4 solutions of the ODE system. 
clear Y0
h = 1e-16;
Y0 = [S0; I0; R0];

gamma_complex = complex(gamma,h);
params = [gamma_complex k delta r];
[~,Y] = ode45(@SIR_rhs,t_vals, Y0, ode_options, params);
S_gamma = imag(Y(:,1))/h; I_gamma = imag(Y(:,2))/h; R_gamma = imag(Y(:,3))/h;


k_complex = complex(k,h);
params = [gamma k_complex delta r];
[~,Y] = ode45(@SIR_rhs,t_vals,Y0,ode_options,params);
S_k = imag(Y(:,1))/h; I_k = imag(Y(:,2))/h; R_k = imag(Y(:,3))/h;
  
delta_complex = complex(delta,h);
params = [gamma k delta_complex r];
[~,Y] = ode45(@SIR_rhs,t_vals,Y0,ode_options,params);
S_delta = imag(Y(:,1))/h; I_delta = imag(Y(:,2))/h; R_delta = imag(Y(:,3))/h;



r_complex = complex(r,h);
params = [gamma k delta r_complex];
[t,Y] = ode45(@SIR_rhs, t_vals,Y0,ode_options,params);
S_r = imag(Y(:,1))/h; I_r = imag(Y(:,2))/h; R_r = imag(Y(:,3))/h;


figure(4)
plot(t,S_gamma_sen','-b',t,S_gamma,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('S_\gamma')
legend('Sensitivity Eq','Complex-Step','Location','SouthEast')


figure(5)
plot(t,I_gamma_sen,'-b',t,I_gamma,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('I_\gamma')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast')

figure(6)
plot(t,R_gamma_sen,'-b',t,R_gamma,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('R_\gamma')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast')


figure(7)
plot(t,S_k_sen,'-b',t,S_k,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('S_k')
legend('Sensitivity Eq','Complex-Step','Location','SouthEast')
%  print -depsc S_r

figure(8)
plot(t,I_k_sen,'-b',t,I_k,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
%  ylabel('$\displaystyle\frac{\partial I}{\partial r}$','interpreter','latex')
ylabel('I_k')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast')

figure(9)
plot(t,R_k_sen,'-b',t,R_k,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
%  ylabel('$\displaystyle\frac{\partial I}{\partial r}$','interpreter','latex')
ylabel('R_k')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast')

figure(10)
plot(t,S_delta_sen,'-b',t,S_delta,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('S_\delta')
legend('Sensitivity Eq','Complex-Step','Location','SouthEast')

figure(11)
plot(t,I_delta_sen','-b',t,I_delta,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('I_\delta')
legend('Sensitivity Eq','Complex-Step','Location','SouthEast')

figure(12)
plot(t,R_delta_sen','-b',t,R_delta,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('R_\delta')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast')

figure(13)
plot(t,S_r_sen,'-b',t,S_r,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
ylabel('S_r')
legend('Sensitivity Eq','Complex-Step','Location','SouthEast')
%  print -depsc S_r

figure(14)
plot(t,I_r_sen,'-b',t,I_r,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
%ylabel('$\displaystyle\frac{\partial I}{\partial r}$','interpreter','latex')
ylabel('I_r')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast') 

figure(15)
plot(t,R_r_sen,'-b',t,R_r,'--r','linewidth',3)
set(gca,'Fontsize',22);
xlabel('Time')
%ylabel('$\displaystyle\frac{\partial I}{\partial r}$','interpreter','latex')
ylabel('R_r')
legend('Sensitivity Eq','Complex-Step','Location','NorthEast') 


 
Sens_mat = [R_gamma R_k R_delta R_r];  
eta=1e-10;   
[Id, UnId] = PSS_eig(Sens_mat,eta);
  
  
function dy = SIR_rhs(~,y,params)
N = 1000;
gamma = params(1); k = params(2); delta = params(3); r = params(4);
S = y(1);          I = y(2);           R = y(3);

dy = [delta*(N-S)-gamma*k*I*S;
      gamma*k*I*S-(r + delta)*I;
      r*I - delta*R];
end



function dy = SIR_sens(~, y,params)
N=1000;
gamma = params(1); k = params(2); delta = params(3); r = params(4);
S = y(1);          I = y(2);           R = y(3);
S_gamma = y(4);    I_gamma = y(5);     R_gamma = y(6);
S_k= y(7);         I_k = y(8);         R_k = y(9);
S_delta = y(10);   I_delta = y(11);    R_delta = y(12);
S_r = y(13);       I_r = y(14);        R_r = y(15);

Svec = [S_gamma; I_gamma; R_gamma; S_k; I_k; R_k; S_delta; I_delta; R_delta;  S_r; I_r; R_r];

J = [-(delta+gamma*k*I)  -gamma*k*S  0; gamma*k*I  (gamma*k*S - (r+delta)) 0; 0  r  -delta];

Der = blkdiag(J,J,J,J); Grad = [-k*I*S; k*I*S; 0; -gamma*I*S;  gamma*I*S; 0; (N-S);  -I;  -R; 0; -I; I];
Sen = Der*Svec + Grad;
dy = [delta*(N-S)-gamma*k*I*S;
      gamma*k*I*S-(r + delta)*I;
      r*I - delta*R;
      Sen];
end


function [Id, UnId] = PSS_eig(Sens_mat, eta)

% p number of parameters.
[~,p]=size(Sens_mat);
%Assume all of the parameters are identifiable.
Id=1:p; 
for k=1:p
    [V, D]=eig(Sens_mat'*Sens_mat);
    [d,ind] = sort(abs(diag(D)));
    Vs = V(:,ind);
    if d(1) > eta
        break
    else
       [~,y]=max(abs(Vs(:,1))); %find the position of maximum element in singular vector that is the last column of V matrix.
       Sens_mat(:,y)= [];%remove the column corresponding to above position 
       Id(y)=[]; % Since y'th element is not identifiable we remove it from the identifiable element subset
    end         
end

UnId=1:p; %Define the subset for the unidentifiable parameters.
UnId(Id)=[]; % Remove all parameters that is identifiable from UnId set.
end

File: Project 1/UQ_8_8.py
"""
UQ Problem 8.8: Compute sensitivities of SIR model and find identifiable subset

This script demonstrates:
1. Computing sensitivities using sensitivity equations
2. Computing sensitivities using complex-step approximation
3. Identifying identifiable parameter subsets using eigenvalue decomposition

SIR Model:
  dS/dt = delta*(N-S) - gamma*k*I*S
  dI/dt = gamma*k*I*S - (r + delta)*I
  dR/dt = r*I - delta*R
"""

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


def sir_rhs(t, y, params):
    """SIR model differential equations."""
    N = 1000
    gamma, k, delta, r = params
    S, I, R = y
    
    dS = delta * (N - S) - gamma * k * I * S
    dI = gamma * k * I * S - (r + delta) * I
    dR = r * I - delta * R
    
    return [dS, dI, dR]


def sir_sensitivity_rhs(t, y, params):
    """SIR model with sensitivity equations."""
    N = 1000
    gamma, k, delta, r = params
    
    # State variables
    S, I, R = y[0], y[1], y[2]
    
    # Sensitivity variables (12 total: 3 states x 4 parameters)
    S_gamma, I_gamma, R_gamma = y[3], y[4], y[5]
    S_k, I_k, R_k = y[6], y[7], y[8]
    S_delta, I_delta, R_delta = y[9], y[10], y[11]
    S_r, I_r, R_r = y[12], y[13], y[14]
    
    # State equations
    dS = delta * (N - S) - gamma * k * I * S
    dI = gamma * k * I * S - (r + delta) * I
    dR = r * I - delta * R
    
    # Jacobian matrix
    J = np.array([
        [-(delta + gamma * k * I), -gamma * k * S, 0],
        [gamma * k * I, gamma * k * S - (r + delta), 0],
        [0, r, -delta]
    ])
    
    # Gradient terms for each parameter
    # Sensitivity w.r.t gamma
    grad_gamma = np.array([-k * I * S, k * I * S, 0])
    sens_gamma = J @ np.array([S_gamma, I_gamma, R_gamma]) + grad_gamma
    
    # Sensitivity w.r.t k
    grad_k = np.array([-gamma * I * S, gamma * I * S, 0])
    sens_k = J @ np.array([S_k, I_k, R_k]) + grad_k
    
    # Sensitivity w.r.t delta
    grad_delta = np.array([N - S, -I, -R])
    sens_delta = J @ np.array([S_delta, I_delta, R_delta]) + grad_delta
    
    # Sensitivity w.r.t r
    grad_r = np.array([0, -I, I])
    sens_r = J @ np.array([S_r, I_r, R_r]) + grad_r
    
    return [dS, dI, dR] + list(sens_gamma) + list(sens_k) + list(sens_delta) + list(sens_r)


def pss_eig(sens_mat, eta):
    """
    Parameter Subset Selection using eigenvalue decomposition.
    
    Returns identifiable and unidentifiable parameter indices.
    """
    n, p = sens_mat.shape
    identifiable = list(range(p))
    sens_work = sens_mat.copy()
    
    for _ in range(p):
        F = sens_work.T @ sens_work
        eigenvalues, eigenvectors = np.linalg.eig(F)
        
        # Sort by eigenvalue magnitude
        idx = np.argsort(np.abs(eigenvalues))
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]
        
        if np.abs(eigenvalues[0]) > eta:
            break
        else:
            # Find parameter with largest contribution to smallest eigenvector
            max_idx = np.argmax(np.abs(eigenvectors[:, 0]))
            sens_work = np.delete(sens_work, max_idx, axis=1)
            identifiable.pop(max_idx)
    
    unidentifiable = [i for i in range(p) if i not in identifiable]
    return identifiable, unidentifiable


def main():
    # Time parameters
    tf = 5
    dt = 0.1
    t_vals = np.arange(0, tf + dt, dt)
    
    # Initial conditions
    S0, I0, R0 = 900, 100, 0
    
    # Parameters
    gamma, k, delta, r = 0.2, 0.1, 0.15, 0.6
    params = [gamma, k, delta, r]
    
    # Initial conditions for sensitivity equations (all zeros)
    y0_sens = [S0, I0, R0] + [0] * 12
    
    # Solve coupled system
    sol_sens = solve_ivp(
        lambda t, y: sir_sensitivity_rhs(t, y, params),
        [0, tf],
        y0_sens,
        t_eval=t_vals,
        rtol=1e-8
    )
    
    # Extract sensitivities from sensitivity equations
    S_gamma_sen, I_gamma_sen, R_gamma_sen = sol_sens.y[3], sol_sens.y[4], sol_sens.y[5]
    S_k_sen, I_k_sen, R_k_sen = sol_sens.y[6], sol_sens.y[7], sol_sens.y[8]
    S_delta_sen, I_delta_sen, R_delta_sen = sol_sens.y[9], sol_sens.y[10], sol_sens.y[11]
    S_r_sen, I_r_sen, R_r_sen = sol_sens.y[12], sol_sens.y[13], sol_sens.y[14]
    
    # Complex-step approximation
    h = 1e-16
    y0 = [S0, I0, R0]
    
    def solve_sir_complex(params_complex):
        sol = solve_ivp(
            lambda t, y: sir_rhs(t, y, params_complex),
            [0, tf],
            y0,
            t_eval=t_vals,
            rtol=1e-8
        )
        return sol.y
    
    # Sensitivity w.r.t gamma
    params_gamma = [complex(gamma, h), k, delta, r]
    Y_gamma = solve_sir_complex(params_gamma)
    S_gamma = np.imag(Y_gamma[0]) / h
    I_gamma = np.imag(Y_gamma[1]) / h
    R_gamma = np.imag(Y_gamma[2]) / h
    
    # Sensitivity w.r.t k
    params_k = [gamma, complex(k, h), delta, r]
    Y_k = solve_sir_complex(params_k)
    S_k = np.imag(Y_k[0]) / h
    I_k = np.imag(Y_k[1]) / h
    R_k = np.imag(Y_k[2]) / h
    
    # Sensitivity w.r.t delta
    params_delta = [gamma, k, complex(delta, h), r]
    Y_delta = solve_sir_complex(params_delta)
    S_delta = np.imag(Y_delta[0]) / h
    I_delta = np.imag(Y_delta[1]) / h
    R_delta = np.imag(Y_delta[2]) / h
    
    # Sensitivity w.r.t r
    params_r = [gamma, k, delta, complex(r, h)]
    Y_r = solve_sir_complex(params_r)
    S_r = np.imag(Y_r[0]) / h
    I_r = np.imag(Y_r[1]) / h
    R_r = np.imag(Y_r[2]) / h
    
    # Create plots comparing sensitivity equations vs complex-step
    fig, axes = plt.subplots(4, 3, figsize=(15, 16))
    param_names = [r'\gamma', 'k', r'\delta', 'r']
    state_names = ['S', 'I', 'R']
    
    sens_eq_data = [
        [S_gamma_sen, I_gamma_sen, R_gamma_sen],
        [S_k_sen, I_k_sen, R_k_sen],
        [S_delta_sen, I_delta_sen, R_delta_sen],
        [S_r_sen, I_r_sen, R_r_sen]
    ]
    
    complex_data = [
        [S_gamma, I_gamma, R_gamma],
        [S_k, I_k, R_k],
        [S_delta, I_delta, R_delta],
        [S_r, I_r, R_r]
    ]
    
    for i, param in enumerate(param_names):
        for j, state in enumerate(state_names):
            ax = axes[i, j]
            ax.plot(t_vals, sens_eq_data[i][j], '-b', linewidth=2, label='Sensitivity Eq')
            ax.plot(t_vals, complex_data[i][j], '--r', linewidth=2, label='Complex-Step')
            ax.set_xlabel('Time')
            ax.set_ylabel(f'${state}_{{{param}}}$')
            ax.legend()
            ax.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('Figures/SIR_sensitivities.png', dpi=150)
    plt.show()
    
    # Parameter subset selection
    sens_mat = np.column_stack([R_gamma, R_k, R_delta, R_r])
    eta = 1e-10
    identifiable, unidentifiable = pss_eig(sens_mat, eta)
    
    print("\n=== Parameter Subset Selection Results ===")
    param_list = ['gamma', 'k', 'delta', 'r']
    print(f"Identifiable parameters: {[param_list[i] for i in identifiable]}")
    print(f"Unidentifiable parameters: {[param_list[i] for i in unidentifiable]}")


if __name__ == "__main__":
    main()
